### DISCUSSION TITLE
Transfering Reinforcement Learning trained policy to real Crazyflie

### DISCUSSION THREAD

**Discussion by piratax007** [bitcraze/discussions#1655] [Q&A]
Hello everybody, 

I'm wondering if someone can help me understand the best approach to transfer a Reinforcement Learning (RL) policy to a real Crazyflie. Currently, I'm working with [Gym-Pybullet Drones](https://github.com/utiasDSL/gym-pybullet-drones) to train an RL policy for low-level control of a Crazyflie - the policy receives position, Euler angles, linear velocity, and angular velocity (all in global frame) as observations and return 4 values between [-1, 1] as actions (RPMs).

To control the drone, the policy actions are mapped to RPMs and then to PWM before being sent to the drone through Crazyradio. I'm using Vicon to capture the state of the Crazyflie and a ROS2 node to build the observation space, compute the actions using the trained neural network, and send the calculated PWMs to the drone using the `setParam` method.

Until now this approach hasn't worked and I have found in some discussions that for RL-based low-level control for Crazyflie, the best approach is to integrate the neural network calculation into the crazyflie firmware. I haven't previous experience modifying the firmware of crazyflie. 

I know that the firmware of crazyflie is free software and can be modified, beyond that, it's possible to integrate the neural network computation inside the firmware? if it is, how can I achieve it? otherwise, does someone have insightful points about how to control a Crazyflie using an RL policy?

Thank you all.

**Comment by gemenerik** [comment]
Communication delays can be a challenge with offboard RL controllers, so integrating the network directly into the firmware is a great way to achieve low-latency control. If your network is small enough, you could use an out-of-tree (OOT) controller implemented as an app. Here’s some helpful documentation: [OOT Controllers](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/oot/#oot-controllers), and a [blog post with an example](https://www.bitcraze.io/2023/02/adding-an-estimator-or-controller/).

You can write your network in C from scratch, but there might be libraries available to make this easier. It’s likely others have done something similar, so you may be able to find additional examples online to guide you.

For larger networks, you might consider using an AI deck. However, this approach introduces communication delays between the deck and the Crazyflie, which may not be ideal for low-level control.

**Reply by piratax007** [reply]
Sorry for the late answer, currently I'm working to implement the NN as an OOT controller.

I'll update the discussion with the news.

Thank you.

**Comment by piratax007** [comment]
Hello again, I have implemented my OOT controller.

Now, I need some help to pass the PWM values, the outputs for the neural network, to the `controller_t`.

As part of the `stabilizar_types.h` interface, I found this:

```c
typedef union {
  uint16_t list[STABILIZER_NR_OF_MOTORS];
  struct {
    uint16_t m1;  // PWM ratio
    uint16_t m2;  // PWM ratio
    uint16_t m3;  // PWM ratio
    uint16_t m4;  // PWM ratio
  } motors;
} motors_thrust_pwm_t;
``` 

but I can't find how to reach it through `controller_t` without modifying its definition.

Thank you for your time.

**Reply by gemenerik** [reply]
Hi, check out [this thread from our old forums](https://forum.bitcraze.io/viewtopic.php?t=5089). Hope that gets you started.
